contracts/src/registry/ETHRegistry.sol

```diff
@@ -1,26 +1,26 @@
 // SPDX-License-Identifier: MIT
 pragma solidity >=0.8.13;
 
-import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
 import {IERC165} from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
 import {ERC1155Singleton} from "./ERC1155Singleton.sol";
 import {IERC1155Singleton} from "./IERC1155Singleton.sol";
 import {IRegistry} from "./IRegistry.sol";
 import {IRegistryDatastore} from "./IRegistryDatastore.sol";
 import {BaseRegistry} from "./BaseRegistry.sol";
 import {PermissionedRegistry} from "./PermissionedRegistry.sol";
+import {EnhancedAccessControl} from "./EnhancedAccessControl.sol";
+import {Roles} from "./Roles.sol";  
 import {IRegistryMetadata} from "./IRegistryMetadata.sol";
 import {MetadataMixin} from "./MetadataMixin.sol";
 import {IETHRegistry} from "./IETHRegistry.sol";
 import {NameUtils} from "../utils/NameUtils.sol";
 
-contract ETHRegistry is PermissionedRegistry, AccessControl, MetadataMixin, IETHRegistry {
-    bytes32 public constant REGISTRAR_ROLE = keccak256("REGISTRAR_ROLE");
 
+contract ETHRegistry is PermissionedRegistry, MetadataMixin, IETHRegistry {
     mapping(uint256 => address) public tokenObservers;
     
     constructor(IRegistryDatastore _datastore, IRegistryMetadata _metadata) PermissionedRegistry(_datastore) MetadataMixin(_metadata) {
-        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRoles(ROOT_RESOURCE, ROLE_REGISTRAR | ROLE_REGISTRAR_ADMIN | ROLE_RENEW | ROLE_RENEW_ADMIN, _msgSender());
```

Comment:
I like the simplicity of using only one access control mechanism for everything. But, mixing access control for our users and access control for system functions (registrar, renewal manager) makes me really nervous.

Granting a user one of these privileges on a non-root resource shouldn't matter, theoretically, but it seems entirely too close to being a vulnerability for comfort.

Thoughts?

Reply:
We could gatekeep things at higher level - e.g by overriding the grant/revoke methods - but I want to keep the lower levels (EAC) clean and pure. This is also why I wanted to put all the roles in the system into a single `Roles.sol` file so that there is no accidental overlap between roles.

Reply:
I'm more wondering if we should use a different ACL mechanism (eg, OZ's) for access rights relating to the contract infrastructure. I have real worries about anything that combines permission checking for system-level privileges like controllers with permission checking for user-owned resources.

Reply:
Ok, let's finalize the acl system first and then think about it.